src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/TagSet.cs

```diff
@@ -0,0 +1,235 @@
+        {
+            case string tag:
+                target[0] = tag;
+                break;
+            case string[] tags:
+                tags.CopyTo(target);
+                break;
+        }
+    }
+
+    internal static TagSet Create(IEnumerable<string>? tags)
+    {
+        if (tags is null)
+        {
+            return Empty;
+        }
+
+        // note that in multi-tag scenarios we always create a defensive copy
+        if (tags is ICollection<string> collection)
+        {
+            switch (collection.Count)
+            {
+                case 0:
+                    return Empty;
+                case 1 when collection is IList<string> list:
+                    return new TagSet(list[0]);
+                case 1:
+                    // avoid the GetEnumerator() alloc
+                    var arr = ArrayPool<string>.Shared.Rent(1);
+                    collection.CopyTo(arr, 0);
+                    string tag = arr[0];
+                    ArrayPool<string>.Shared.Return(arr);
+                    return new TagSet(tag);
+                default:
+                    arr = new string[collection.Count];
+                    collection.CopyTo(arr, 0);
+                    return new TagSet(arr);
+            }
+        }
+
+        // perhaps overkill, but: avoid as much as possible when unrolling
+        using var iterator = tags.GetEnumerator();
+        if (!iterator.MoveNext())
+        {
+            return Empty;
+        }
+
+        var firstTag = iterator.Current;
+        if (!iterator.MoveNext())
+        {
+            return new TagSet(firstTag);
+        }
+
+        string[] oversized = ArrayPool<string>.Shared.Rent(8);
+        oversized[0] = firstTag;
+        int count = 1;
+        do
+        {
+            if (count == oversized.Length)
+            {
+                // grow
+                var bigger = ArrayPool<string>.Shared.Rent(count * 2);
+                oversized.CopyTo(bigger, 0);
+                ArrayPool<string>.Shared.Return(oversized);
+                oversized = bigger;
+            }
+
+            oversized[count++] = iterator.Current;
+        }
+        while (iterator.MoveNext());
+
+        if (count == oversized.Length)
+        {
+            return new TagSet(oversized);
+        }
+        else
+        {
+            var final = oversized.AsSpan(0, count).ToArray();
+            ArrayPool<string>.Shared.Return(oversized);
+            return new TagSet(final);
+        }
+    }
+
+    internal string[] ToArray() // for testing only
+    {
+        var arr = new string[Count];
+        CopyTo(arr);
+        return arr;
+    }
+
+    internal const string WildcardTag = "*";
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1122:Use string.Empty for empty strings", Justification = "Not needed")]
+    internal bool TryFind(ReadOnlySpan<char> span, [NotNullWhen(true)] out string? tag)
+    {
+        switch (_tagOrTags)
```

Comment:
No `null` or `default` case?

Reply:
not needed - it would be `tag = null; return false;` which is what already happens *after* the `switch` (sharing the "nope" logic from all ways of saying "nope")